### PR TITLE
[Core] Damage Mitigation Tab

### PR DESCRIPTION
**Pull request by Gebuz**
Early start on #1083. Splits all damage mitigated (excluding asorbs and blocks) evenly between all damage reductions using the formula explained in the issue.

#### Implemented
- All buffs (except Devo & Aura of Sac)
- All debuffs
- Supports both stacking buffs/debuffs and multiple instances (double Ironbark for instance)
- most passives, including armor and versatility.
- Fully absorbed blocks. They do not have a `blocked` property, and it aint likely to be added any time soon. I don't think it's possible to see if the combatant has a shield equipped, so for now it just assigns any unknown mitigation on fully absorbed hits for Prot pala & warror as block.
- Support for Avoidance/Feint. Still needs a complete whitelist to function properly.
- Mastery for Protection Paladin.
- Added armor tracker.
- Added Aura of Sac and Devo Aura.

#### Missing
- Avoidance. Needs a complete whitelist to function properly.
- Mastery for Destruction Warlock. Destro is impossible because it's random, so will just assign any unknown mitigation (up to the mastery maximum) to it.
- Write some text explaining the calculations (Link to Zero's article).

#### Screenshots
image
image
image

### CODE REVIEW

**Comment by Gebuz** [comment]
Added a way to assign any unknown damage mitigation to Aura of Sacrifice and Devotion Aura. It doesn't quite work with multiple instances of the same aura yet (currently it just adds it as a single value so it gets a smaller portion of dmg mitigation assigned than it should).
image

**Comment by Gebuz** [comment]
I'm not sure how much more time I'll have to work on this before BFA release, and I likely wont be working on it at all between that and clearing mythic Uldir. I think it's in an okay state for the most part, but it might not be complete enough that we want to show it yet.

If it's not in an acceptable state I will close it for now and come back to it when I find the time. If anybody else wants to pick it up in the mean time let me know you're welcome to, just PM me on discord.

It is currently missing the following:
#### Major
- Whitelist of AOE abilities that is reduced by avoidance (implementation is ready, but it only contains Garothi Worldbreaker abilities at the moment).
- Update stacktracker to include all temporary avoidance and versatility buffs.
- Destruction Mastery. Could be implemented to some extend by guessing, but it is impossible to tell this and other unknown sources like Devotion Aura apart.
- Spirit Link Totem's buff does not show up at all. Could contribute any ~10% unknown DR to it, but just like the above it's hard with multiple unknown sources.

#### Minor
 - Fully absorbed hits does not have a block event. Currently any unknown is contributed to block when this happens, but it might be a good idea to at least contribute the minimum value to devotion aura.
 - Debuffs fade before death events, causing the killing blow to look unmitigated.
 - Perseverence also triggers from other priests Renew if the combatant has the talent.
 - I don't know how Rugged Tenacity works yet. I'd assume it works like block, but the value is so low it mostly doesn't matter anyway.
 - The unknown mitigation assignement to Auras could be improved for the case where the paladin with Devotion Aura is on his own, or when multiple of the same aura are present and AM is active for at least one of them.

File: src/Parser/Warrior/Fury/Modules/Talents/FuriousSlashTimesByStacks.js [hunk 1]
```diff
@@ -1,63 +1,63 @@
...
-			return;
-		}
-		
-		this.furiousSlashStacks[this.lastFuriousSlashStack].push(event.timestamp - this.lastFuriousSlashUpdate);
-		this.lastFuriousSlashUpdate = event.timestamp;
-		this.lastFuriousSlashStack = 0;
-	}
-}
-
+import Analyzer from 'Parser/Core/Analyzer';
```
Comment (MartijnHols) [inline src/Parser/Warrior/Fury/Modules/Talents/FuriousSlashTimesByStacks.js]: What did you change about this module? I can't see it because everything is marked as changed.

File: src/Parser/Core/Modules/EnemyInstances.js [hunk 2]
```diff
@@ -5,11 +5,11 @@ const debug = true;
 
 class EnemyInstances extends Enemies {
   getEntity(event) {
-    if (event.targetIsFriendly) {
+    if (event.targetIsFriendly && event.sourceIsFriendly) {
```
Comment (MartijnHols) [inline src/Parser/Core/Modules/EnemyInstances.js]: Why did you change this? I don't think this is supposed to include damage to friendly targets (e.g. other players or player pets), but your change would lead to including friendly targets given the source isn't friendly (e.g. another player hit by the boss).

File: src/Parser/Core/Modules/DamageMitigation/Reductions.js [hunk 3]
```diff
@@ -0,0 +1,449 @@
+import SPELLS from 'common/SPELLS';
+import MAGIC_SCHOOLS from 'common/MAGIC_SCHOOLS';
+import SPECS from 'Game/SPECS';
+import RACES from 'Game/RACES';
+
+import AOE_TYPE from './AOE_TYPE';
+
```
Comment (MartijnHols) [inline src/Parser/Core/Modules/DamageMitigation/Reductions.js]: I'd prefer it if this file didn't include spells that are class/spec specific. I believe core shouldn't know about class/spec specific things unless they can be applied to others (e.g. externals).

File: src/Parser/Core/Modules/DamageMitigation/DamageMitigation.js [hunk 4]
```diff
@@ -0,0 +1,380 @@
...
+      }
+      if (unknown < - BUFFER_PERCENT) {
+        debug && console.warn('The actual percent mitigated was lower than expected given the mitigations active at the time. Actual: ' + formatPercentage(percentMitigated) + '%, Expected: ' + formatPercentage(multiplicative) + '%');
+      }
+    }
+
+    // If the actual percent mitigated is higher than expected, try to assign the unknown mitigation.
+    if (unknown > BUFFER_PERCENT) {
+      // If the player is using a shield and the hit was fully absorbed, assume the remaining mitigation was block (Block amount is not in the log when the hit fully absorbed).
+      if (event.amount === 0 && (this.selectedCombatant.specId === SPECS.PROTECTION_PALADIN.id ||
```
Comment (MartijnHols) [inline src/Parser/Core/Modules/DamageMitigation/DamageMitigation.js]: I'd prefer it if this file didn't include effects that are class/spec specific. I believe core shouldn't know about class/spec specific things unless they can be applied to others (e.g. externals).

File: src/Parser/Core/Modules/DamageMitigation/AOE_ABILITIES.js [hunk 5]
```diff
@@ -0,0 +1,50 @@
+// White list of all AOE abilities (INCOMPLETE!)
+export default 
+[
+  // Garothi World Breaker
+  {
+    id: 244761,
```
Comment (MartijnHols) [inline src/Parser/Core/Modules/DamageMitigation/AOE_ABILITIES.js]: Please include the spell name in a comment.

File: src/Parser/Core/Modules/DamageMitigation/AOE_TYPE.js [hunk 6]
```diff
@@ -0,0 +1,5 @@
+export default {
+  BOTH: 'Both',
+  AOE_ONLY: 'AOE only',
+  NON_AOE_ONLY: 'Non AOE only',
```
Comment (MartijnHols) [inline src/Parser/Core/Modules/DamageMitigation/AOE_TYPE.js]: How about `DIRECT_ONLY`?

File: src/Interface/Others/DamageMitigationBreakdown.js [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+    const mitigated = this.prepareMitigated(tracker.mitigated);
+
+    let totalMitigated = tracker.totalMitigated;
+    totalMitigated = (totalMitigated === 0) ? 1 : totalMitigated;
+
+    return (
+      <div>
+        {!tracker.isAccurate && (
+          <Danger style={{ marginTop: 10, marginBottom: 10 }}>
+            There is a large amount of damage mitigation which's source is unknown. This is likely caused by our avoidance or versatility tracker being incomplete, our list of abilities affected by avoidance being incomplete or one of the following abilities being partially untrackable in the log, and we can only guess at how much it actually mitigated: Devotion Aura, Aura of Sacrifice, Destruction Warlock Mastery and Spirit Link Totem.
```
Comment (MartijnHols) [inline src/Interface/Others/DamageMitigationBreakdown.js]: The usage of the word "which's" makes this somewhat hard to read

File: src/Parser/Core/Modules/DamageMitigation/Armor.js [hunk 8]
```diff
@@ -0,0 +1,24 @@
+import Analyzer from 'Parser/Core/Analyzer';
+
+class Armor extends Analyzer {
```
Comment (MartijnHols) [inline src/Parser/Core/Modules/DamageMitigation/Armor.js]: Why did you make this class instead of using StatTracker?

File: src/Parser/Core/Modules/DamageMitigation/DamageMitigation.js [hunk 9]
```diff
@@ -0,0 +1,380 @@
...
+  initReduction(reduction) {
+    this.mitigated[reduction.id] = {
+      name: reduction.name,
+      amount: 0,
+    };
+  }
+
+  getMitigation(reduction, event) {
+    // Calculate Dampen Harm reduction. 
+    if (reduction.id === SPELLS.DAMPEN_HARM_TALENT.id) {
```
Comment (MartijnHols) [inline src/Parser/Core/Modules/DamageMitigation/DamageMitigation.js]: I'd prefer it if this file didn't include effects that are class/spec specific. I believe core shouldn't know about class/spec specific things unless they can be applied to others (e.g. externals).

File: src/Parser/Core/Modules/DamageMitigation/Reductions.js (hunk 3 above)
Comment (Gebuz) [reply src/Parser/Core/Modules/DamageMitigation/Reductions.js]: Agreed, dampen harm and the fully absorbed block should be moved to a class specific extension.

File: src/Parser/Core/Modules/DamageMitigation/AOE_ABILITIES.js (hunk 5 above)
Comment (Gebuz) [reply src/Parser/Core/Modules/DamageMitigation/AOE_ABILITIES.js]: Will do in the future, was just being lazy with the test implementation. On that note, I later realized that there is a good chance that ticks are never reduced by avoidance, in which case the whitelist could simply be an array of ids. I never got around to test that theory though.

File: src/Parser/Core/Modules/DamageMitigation/Armor.js (hunk 8 above)
Comment (Gebuz) [reply src/Parser/Core/Modules/DamageMitigation/Armor.js]: Using the armor from the damage event is more accurate. Could update the statTracker to be updated based on that, but it might be confusing that the implementation is so different to the other stats. Also it would not update until a damage event, so if for some reason you tried to get the armor for something else (idk what that would be) between hits it might not return the correct value.

File: src/Parser/Warrior/Fury/Modules/Talents/FuriousSlashTimesByStacks.js (hunk 1 above)
Comment (MartijnHols) [reply src/Parser/Warrior/Fury/Modules/Talents/FuriousSlashTimesByStacks.js]: Please try updating from master

**Comment by Gebuz** [comment]
Closing for now, will return to it when I'm not spending all my free time playing the game (so likely after uldir mythic progress is over).
